File: api_sample/python/cdb-oauth2-list-systems/test_cdb_oauth2_sample.py
# Copyright 2018-present Network Optix, Inc. Licensed under MPL 2.0: www.mozilla.org/MPL/2.0/
"""
Offline tests for cdb_oauth2_sample.py. No network, no account needed.

Run from this folder:  pytest -v
"""

import argparse

import pytest

import cdb_oauth2_sample as sample


# ---------------------------------------------------------------------------
# Test doubles
# ---------------------------------------------------------------------------

class FakeResponse:
    def __init__(self, status_code=200, json_data=None, text=""):
        self.status_code = status_code
        self._json = json_data
        self.text = text

    @property
    def ok(self):
        return self.status_code < 400

    def json(self):
        if self._json is None:
            raise ValueError("no json")
        return self._json


class FakeSession:
    """Serves a queued response per HTTP verb and records what was sent."""

    def __init__(self, post=None, get=None):
        self.verify = None
        self._post = post
        self._get = get
        self.post_url = self.post_json = None
        self.get_url = self.get_headers = None

    def post(self, url, json=None, timeout=None):
        self.post_url, self.post_json = url, json
        return self._post

    def get(self, url, headers=None, timeout=None):
        self.get_url, self.get_headers = url, headers
        return self._get


# ---------------------------------------------------------------------------
# login()
# ---------------------------------------------------------------------------

def test_login_returns_and_stores_token():
    session = FakeSession(post=FakeResponse(200, {"access_token": "nxcdb-xyz"}))
    client = sample.NxCloudOAuthClient("https://nxvms.com", "me@x.com", "pw",
                                       session=session)

    token = client.login()

    assert token == "nxcdb-xyz"
    assert client.token == "nxcdb-xyz"
    assert session.post_url == "https://nxvms.com/cdb/oauth2/token"
    # The documented password-grant fields must be present.
    assert session.post_json["grant_type"] == "password"
    assert session.post_json["client_id"] == "3rdParty"
    assert session.post_json["username"] == "me@x.com"
    # No 2FA code was supplied, so it must not be in the body.
    assert "mfaCode" not in session.post_json


def test_login_includes_mfa_code_when_set():
    session = FakeSession(post=FakeResponse(200, {"access_token": "t"}))
    client = sample.NxCloudOAuthClient("https://nxvms.com", "me@x.com", "pw",
                                       mfa_code="123456", session=session)
    client.login()
    assert session.post_json["mfaCode"] == "123456"


def test_login_bad_credentials_raises():
    session = FakeSession(post=FakeResponse(401, text="no"))
    client = sample.NxCloudOAuthClient("https://nxvms.com", "me@x.com", "pw",
                                       session=session)
    with pytest.raises(sample.AuthError):
        client.login()


def test_login_missing_token_raises_apierror():
    session = FakeSession(post=FakeResponse(200, {"something_else": 1}))
    client = sample.NxCloudOAuthClient("https://nxvms.com", "me@x.com", "pw",
                                       session=session)
    with pytest.raises(sample.ApiError):
        client.login()


# ---------------------------------------------------------------------------
# list_systems()
# ---------------------------------------------------------------------------

def test_list_systems_sends_bearer_header():
    payload = [{"id": "s1", "name": "HQ", "status": "activated", "version": "6.0"}]
    session = FakeSession(get=FakeResponse(200, payload))
    client = sample.NxCloudOAuthClient("https://nxvms.com", "me@x.com", "pw",
                                       session=session)
    client.token = "nxcdb-abc"  # pretend we already logged in

    sites = client.list_systems()

    assert sites[0]["name"] == "HQ"
    assert session.get_url == "https://nxvms.com/cdb/systems"
    assert session.get_headers["Authorization"] == "Bearer nxcdb-abc"


def test_list_systems_without_login_raises():
    client = sample.NxCloudOAuthClient("https://nxvms.com", "me@x.com", "pw",
                                       session=FakeSession())
    with pytest.raises(sample.ApiError):
        client.list_systems()


# ---------------------------------------------------------------------------
# config + table
# ---------------------------------------------------------------------------

def _args(**overrides):
    base = dict(host=None, user=None, password=None, mfa_code=None,
                cloud_site_id=None)
    base.update(overrides)
    return argparse.Namespace(**base)


def test_cli_overrides_env(monkeypatch):
    monkeypatch.setenv("NX_CLOUD_HOST", "https://env")
    config = sample.resolve_config(_args(host="https://cli"),
                                   {"NX_CLOUD_HOST": "https://file"})
    assert config["host"] == "https://cli"


def test_no_scope_means_cloud_wide_token():
    # Without a cloud_site_id, the request body must NOT carry a scope:
    # that yields a cloud-wide (cdb) token, correct for listing Sites.
    session = FakeSession(post=FakeResponse(200, {"access_token": "t"}))
    client = sample.NxCloudOAuthClient("https://nxvms.com", "me@x.com", "pw",
                                       session=session)
    client.login()
    assert "scope" not in session.post_json


def test_scope_set_when_cloud_site_id_given():
    # With a cloud_site_id, the body must carry scope=cloudSystemId=<id>,
    # which produces a site-scoped token.
    session = FakeSession(post=FakeResponse(200, {"access_token": "t"}))
    client = sample.NxCloudOAuthClient("https://nxvms.com", "me@x.com", "pw",
                                       cloud_site_id="sys-123", session=session)
    client.login()
    assert session.post_json["scope"] == "cloudSystemId=sys-123"


def test_list_systems_unwraps_object_envelope():
    # The real CDB may wrap the array in an object; we must still find sites.
    payload = {"sites": [{"id": "s1", "name": "HQ"}, {"id": "s2", "name": "Lab"}]}
    session = FakeSession(get=FakeResponse(200, payload))
    client = sample.NxCloudOAuthClient("https://nxvms.com", "me@x.com", "pw",
                                       session=session)
    client.token = "nxcdb-abc"
    sites = client.list_systems()
    assert [s["name"] for s in sites] == ["HQ", "Lab"]


def test_extract_systems_handles_shapes():
    bare = [{"id": "s1"}]
    assert sample.extract_systems(bare) is bare
    assert sample.extract_systems({"sites": bare}) == bare
    assert sample.extract_systems({"reply": bare}) == bare
    assert sample.extract_systems({"data": {"sites": bare}}) == bare
    # Unknown key but still a list of objects -> found by the fallback scan.
    assert sample.extract_systems({"whatever": bare}) == bare
    # Genuinely nothing -> empty list.
    assert sample.extract_systems({"count": 0}) == []
    assert sample.extract_systems("nope") == []


def test_format_systems_table():
    out = sample.format_systems_table(
        [{"id": "s1", "name": "HQ", "status": "activated", "version": "6.0"}])
    assert "NAME" in out and "HQ" in out
    assert "No Sites" in sample.format_systems_table([])

File: api_sample/typescript/cdb-oauth2-list-systems/test_cdb_oauth2_sample.ts
// Copyright 2018-present Network Optix, Inc. Licensed under MPL 2.0: www.mozilla.org/MPL/2.0/
/**
 * Offline tests for cdb_oauth2_sample.ts. No network, no account needed.
 *
 * Run from this folder:  node --test test_cdb_oauth2_sample.ts
 */

import assert from "node:assert/strict";
import test from "node:test";

import type { FetchImpl, Site } from "../nx-types.ts";
import {
  NxCloudOAuthClient,
  extractSystems,
  formatSystemsTable,
  resolveConfig,
  AuthError,
  ApiError,
} from "./cdb_oauth2_sample.ts";
import type { CliArgs } from "./cdb_oauth2_sample.ts";

interface FakeResponseInit {
  status?: number;
  json?: unknown;
  text?: string;
}

function makeResponse({ status = 200, json = null, text = "" }: FakeResponseInit = {}): Response {
  return {
    status,
    ok: status < 400,
    async json() {
      if (json === null) throw new Error("no json");
      return json;
    },
    async text() {
      return text;
    },
  } as unknown as Response;
}

interface RecordedCalls {
  post: { url: string; body: Record<string, unknown> | null } | null;
  get: { url: string; headers: Record<string, string> } | null;
}

interface FakeFetch extends FetchImpl {
  calls: RecordedCalls;
}

interface FakeFetchOpts {
  post?: Response | null;
  get?: Response | null;
}

/** Serves one queued response per verb; records what was sent. */
function fakeFetch({ post = null, get = null }: FakeFetchOpts = {}): FakeFetch {
  const calls: RecordedCalls = { post: null, get: null };
  const impl = (async (url: string | URL, options: RequestInit = {}) => {
    if ((options.method || "GET").toString().toUpperCase() === "POST") {
      calls.post = {
        url: String(url),
        body: options.body ? JSON.parse(options.body as string) : null,
      };
      return post;
    }
    calls.get = {
      url: String(url),
      headers: options.headers as Record<string, string>,
    };
    return get;
  }) as unknown as FakeFetch;
  impl.calls = calls;
  return impl;
}

interface MakeClientOpts {
  mfaCode?: string | null;
  cloudSystemId?: string | null;
}

function makeClient(opts: MakeClientOpts = {}, fetchOpts: FakeFetchOpts = {}): {
  client: NxCloudOAuthClient;
  f: FakeFetch;
} {
  const f = fakeFetch(fetchOpts);
  const client = new NxCloudOAuthClient("https://nxvms.com", "me@x.com", "pw", {
    fetchImpl: f,
    ...opts,
  });
  return { client, f };
}

// ---------------------------------------------------------------------------
// login()
// ---------------------------------------------------------------------------

test("login returns and stores the token", async () => {
  const { client, f } = makeClient({}, { post: makeResponse({ json: { access_token: "nxcdb-xyz" } }) });
  const token = await client.login();
  assert.equal(token, "nxcdb-xyz");
  assert.equal(client.token, "nxcdb-xyz");
  assert.equal(f.calls.post!.url, "https://nxvms.com/cdb/oauth2/token");
  assert.equal(f.calls.post!.body!.grant_type, "password");
  assert.equal(f.calls.post!.body!.client_id, "3rdParty");
  assert.equal(f.calls.post!.body!.username, "me@x.com");
  assert.ok(!("mfaCode" in f.calls.post!.body!));
});

test("login includes the mfa code when set", async () => {
  const { client, f } = makeClient(
    { mfaCode: "123456" },
    { post: makeResponse({ json: { access_token: "t" } }) },
  );
  await client.login();
  assert.equal(f.calls.post!.body!.mfaCode, "123456");
});

test("login with bad credentials raises AuthError", async () => {
  const { client } = makeClient({}, { post: makeResponse({ status: 401, text: "no" }) });
  await assert.rejects(() => client.login(), AuthError);
});

test("login without a token raises ApiError", async () => {
  const { client } = makeClient({}, { post: makeResponse({ json: { something_else: 1 } }) });
  await assert.rejects(() => client.login(), ApiError);
});

// ---------------------------------------------------------------------------
// listSystems()
// ---------------------------------------------------------------------------

test("listSystems sends the bearer header", async () => {
  const payload = [{ id: "s1", name: "HQ", status: "activated", version: "6.0" }];
  const { client, f } = makeClient({}, { get: makeResponse({ json: payload }) });
  client.token = "nxcdb-abc";
  const sites = await client.listSystems();
  assert.equal(sites[0]!.name, "HQ");
  assert.equal(f.calls.get!.url, "https://nxvms.com/cdb/systems");
  assert.equal(f.calls.get!.headers.Authorization, "Bearer nxcdb-abc");
});

test("listSystems without login raises", async () => {
  const { client } = makeClient();
  await assert.rejects(() => client.listSystems(), ApiError);
});

test("listSystems unwraps an object envelope", async () => {
  const payload = { sites: [{ id: "s1", name: "HQ" }, { id: "s2", name: "Lab" }] };
  const { client } = makeClient({}, { get: makeResponse({ json: payload }) });
  client.token = "nxcdb-abc";
  const sites = await client.listSystems();
  assert.deepEqual(sites.map((s) => s.name), ["HQ", "Lab"]);
});

// ---------------------------------------------------------------------------
// scope, config, envelope helper, table
// ---------------------------------------------------------------------------

test("no cloud-site-id means no scope (cloud-wide token)", async () => {
  const { client, f } = makeClient({}, { post: makeResponse({ json: { access_token: "t" } }) });
  await client.login();
  assert.ok(!("scope" in f.calls.post!.body!));
});

test("scope is set when a cloud-site-id is given", async () => {
  const { client, f } = makeClient(
    { cloudSystemId: "sys-123" },
    { post: makeResponse({ json: { access_token: "t" } }) },
  );
  await client.login();
  assert.equal(f.calls.post!.body!.scope, "cloudSystemId=sys-123");
});

test("CLI flag overrides env var", () => {
  const args: CliArgs = {
    host: "https://cli",
    user: null,
    password: null,
    mfaCode: null,
    cloudSystemId: null,
    envFile: ".env",
    insecure: false,
    debug: false,
  };
  const config = resolveConfig(args, { NX_CLOUD_HOST: "https://file" }, { NX_CLOUD_HOST: "https://env" });
  assert.equal(config.host, "https://cli");
});

test("extractSystems handles the shapes", () => {
  const bare: Site[] = [{ id: "s1" }];
  assert.equal(extractSystems(bare), bare);
  assert.deepEqual(extractSystems({ sites: bare }), bare);
  assert.deepEqual(extractSystems({ reply: bare }), bare);
  assert.deepEqual(extractSystems({ data: { sites: bare } }), bare);
  assert.deepEqual(extractSystems({ whatever: bare }), bare); // fallback scan
  assert.deepEqual(extractSystems({ count: 0 }), []);
  assert.deepEqual(extractSystems("nope"), []);
});

test("formatSystemsTable renders rows and the empty case", () => {
  const out = formatSystemsTable([{ id: "s1", name: "HQ", status: "activated", version: "6.0" }]);
  assert.ok(out.includes("NAME") && out.includes("HQ"));
  assert.ok(formatSystemsTable([]).includes("No Sites"));
});
